pageserver/src/metrics.rs

```diff
@@ -0,0 +1,418 @@
+use metrics::core::{AtomicU64, GenericCounter};
+use metrics::{
+    register_histogram, register_histogram_vec, register_int_counter, register_int_counter_vec,
+    register_int_gauge, register_int_gauge_vec, register_uint_gauge_vec, Histogram, HistogramVec,
+    IntCounter, IntCounterVec, IntGauge, IntGaugeVec, UIntGauge, UIntGaugeVec,
+};
+use once_cell::sync::Lazy;
+
+/// Prometheus histogram buckets (in seconds) that capture the majority of
+/// latencies in the microsecond range but also extend far enough up to distinguish
+/// "bad" from "really bad".
+fn get_buckets_for_critical_operations() -> Vec<f64> {
+    let buckets_per_digit = 5;
+    let min_exponent = -6;
+    let max_exponent = 2;
+
+    let mut buckets = vec![];
+    // Compute 10^(exp / buckets_per_digit) instead of 10^(1/buckets_per_digit)^exp
+    // because it's more numerically stable and doesn't result in numbers like 9.999999
+    for exp in (min_exponent * buckets_per_digit)..=(max_exponent * buckets_per_digit) {
+        buckets.push(10_f64.powf(exp as f64 / buckets_per_digit as f64))
+    }
+    buckets
+}
+
+// Metrics collected on operations on the storage repository.
+const STORAGE_TIME_OPERATIONS: &[&str] = &[
+    "layer flush",
+    "compact",
+    "create images",
+    "init logical size",
+    "load layer map",
+    "gc",
+];
+
+pub static STORAGE_TIME: Lazy<HistogramVec> = Lazy::new(|| {
+    register_histogram_vec!(
```

Comment:
Not strictly related to this PR, more like an idea for the future.

How about splitting STORAGE_TIME metrics into global and per-timeline? Global metrics can be represented with a single histogram with detailed buckets as it now, while per-timeline metrics can live with only 2 timeseries: count and sum.

Motivation: histogram takes a lot of space, but doesn't provide enough details for a single timeline. We have 6 operations, ~44 exponents and that makes `6 * 44 = 264` timeseries for a single timeline, which is a lot for a single per-timeline metric. By not providing enough details I mean that we still don't know exact individual durations, and I think that it makes more sense to log these durations explicitly for a single timeline, if we want to analyze it.

The same actually applies to most of the other per-timeline histograms.

Reply:
Thanks, sounds like it makes a lot of sense. I guess it makes sense to tackle that afterwards to avoid scope creep. Also probably makes nicer-to-review-PR.